quickwit-common/src/uri.rs

```diff
@@ -18,46 +18,93 @@
 // along with this program. If not, see <http://www.gnu.org/licenses/>.
 
 use std::env;
+use std::fmt::Display;
 use std::path::{Component, Path, PathBuf};
 
-use anyhow::Context;
+use anyhow::{bail, Context};
 
 /// Default file protocol `file://`
 const FILE_PROTOCOL: &str = "file";
 
-/// Normalizes a URI
-/// A `file://` protocol is assumed if not specified.
-/// File URIs are resolved (normalised) relative to the current working directory
-/// unless an absolute path is specified.
-/// Handles special characters like (~, ., ..)
-pub fn normalize_uri(uri: &str) -> anyhow::Result<String> {
-    let (protocol, mut path) = match uri.split_once("://") {
-        None => (FILE_PROTOCOL, uri.to_string()),
-        Some((protocol, path)) => (protocol, path.to_string()),
-    };
-
-    let current_dir = env::current_dir().context("Could not fetch the current directory.")?;
-
-    if protocol == FILE_PROTOCOL {
-        if path.starts_with('~') {
-            let home_dir_path = home::home_dir()
-                .context("Could not fetch the home directory.")?
+const PROTOCOL_SEPARATOR: &str = "://";
+
+/// Encapsulates the URI type.
+#[derive(Debug, PartialEq, Eq)]
+pub struct Uri {
+    uri: String,
+    protocol_idx: usize,
+}
+
+impl Uri {
+    /// Tries to to construct a Uri from the raw string.
+    /// A `file://` protocol is assumed if not specified.
+    /// File URIs are resolved (normalised) relative to the current working directory
+    /// unless an absolute path is specified.
+    /// Handles special characters like (~, ., ..)
+    pub fn try_new(uri: &str) -> anyhow::Result<Self> {
+        let (protocol, mut path) = match uri.split_once(PROTOCOL_SEPARATOR) {
+            None => (FILE_PROTOCOL, uri.to_string()),
+            Some((protocol, path)) => (protocol, path.to_string()),
+        };
+
+        if protocol == FILE_PROTOCOL {
+            let current_dir =
+                env::current_dir().context("Could not fetch the current directory.")?;
+
+            if path.starts_with('~') {
+                if path != "~" && !path.starts_with("~/") {
+                    bail!("This path syntax `{}` is not supported.", uri);
+                }
```

Comment:
I'm struggling to understand the condition.

Here we need to sastify 3 conditions to raise the error:
- `path.starts_with('~')`
- `path != "~"`
- `!path.starts_with("~/")`

But I don't have a better idea without using a regex.... can you add a comment just to explain the condition?

Reply:
```rust
// We only accept `~` and `~/` which are aliases to the home directory.
// If there is something following `~` and it's not `/`, we bail out.
if path.chars()
    .nth(1)
    .map(|second_character| second_character != '/')
    .unwrap_or(false) {
    bail!("This path syntax `{}` is not supported.", uri);
}
```